### ISSUE TITLE
Multiple test failures in Gentoo's network sandbox

### ISSUE THREAD

**Issue by hololeap** [haskell/haskell-language-server#3126]
I am getting a lot of test failures from the `hls-eval-plugin` test suite using 55d90241440c5282b9e60cf80a98fd8698e1f8d9:

```
62 out of 65 tests failed (272.92s)
```

Full build log is [here](https://gist.github.com/hololeap/1d139322b305e007f4e12977051ccfbc#file-build-log-L364-L1254).

### Your environment

GHC 9.0.2 on Gentoo Linux. All packages are installed through the system package manager.

I am building the following packages from 55d90241440c5282b9e60cf80a98fd8698e1f8d9:

- `ghcide`
- `haskell-language-server`
- `hls-alternate-number-format-plugin`
- `hls-change-type-signature-plugin`
- `hls-code-range-plugin`
- `hls-eval-plugin`
- `hls-floskell-plugin`
- `hls-fourmolu-plugin`
- `hls-graph`
- `hls-hlint-plugin`
- `hls-plugin-api`
- `hls-rename-plugin`
- `hls-retrie-plugin`
- `hls-test-utils`

I am building the rest from Hackage with the following versions (let me know if you need more):

- `aeson-2.0.3.0`
- `data-default-0.7.1.1`
- `diff-0.4.0`
- `dlist-1.0`
- `extra-1.7.10`
- `ghc-paths-0.1.0.12`
- `hashable-1.4.0.2`
- `lens-5.1.1`
- `lsp-1.5.0.0`
- `lsp-types-1.5.0.0`
- `megaparsec-9.2.1`
- `parser-combinators-1.3.0`
- `pretty-simple-4.1.1.0`
- `quickcheck-2.14.2`
- `safe-exceptions-0.1.7.1`
- `unliftio-0.2.18`
- `unordered-containers-0.2.17.0`

**Comment by hololeap** [comment]
Are these tests passing elsewhere? If so, can someone give me a link to where they are passing?

**Comment by pepeiborra** [comment]
They are passing in the Github CI. [link](https://github.com/haskell/haskell-language-server/runs/8156445452?check_suite_focus=true#step:17:1)

**Comment by wz1000** [comment]
Perhaps you want to test using the environment variables `LSP_TEST_LOG_STDERR=1` and `LSP_TEST_LOG_MESSAGES=1` to get more information on why it fails.

**Comment by hololeap** [comment]
@wz1000, thanks for the hint about the environment variables.

It looks like this has something to do with Gentoo's build sandboxing. All the tests pass when run outside the sandbox.

I am seeing blocked network calls being logged. It looks like `cabal` is being run as a part of the test, which is trying to download something.

```json
{
    "jsonrpc": "2.0",
    "method": "textDocument/publishDiagnostics",
    "params": {
        "diagnostics": [
            {
                "message": "Error when calling cabal exec -v0 -- ghc --print-libdir\n\ndieVerbatim: user error (cabal: '/usr/bin/curl' exited with an error:\ncurl: (6) Could not resolve host: hackage.haskell.org\n)\n\n",
                "range": {
                    "end": {
                        "character": 0,
                        "line": 1
                    },
                    "start": {
                        "character": 0,
                        "line": 0
                    }
                },
                "severity": 1,
                "source": "cradle"
            }
        ],
        "uri": "file:///var/tmp/portage/dev-haskell/hls-eval-plugin-1.3.0.0/work/hls-eval-plugin-1.3.0.0/test/testdata/TLocalImport.hs",
        "version": 0
    }
}
```

**Comment by hololeap** [comment]
It looks like I'm running into the same thing for `hls-module-rename-plugin-1.1.0.0`:

```json
{
    "jsonrpc": "2.0",
    "method": "textDocument/publishDiagnostics",
    "params": {
        "diagnostics": [
            {
                "message": "Error when calling cabal exec -v0 -- ghc --print-libdir\n\ndieVerbatim: user error (cabal: '/usr/bin/curl' exited with an error:\ncurl: (6) Could not resolve host: hackage.haskell.org\n)\n\n",
                "range": {
                    "end": {
                        "character": 0,
                        "line": 1
                    },
                    "start": {
                        "character": 0,
                        "line": 0
                    }
                },
                "severity": 1,
                "source": "cradle"
            }
        ],
        "uri": "file:///var/tmp/portage/dev-haskell/hls-module-name-plugin-1.1.0.0/work/hls-module-name-plugin-1.1.0.0/test/testdata/canonicalize/Lib/A.hs",
        "version": 0
    }
}
```

```
FAIL (16.59s)
    Received an unexpected message from the server:
    Was parsing: Pattern match failure in do expression at test/Main.hs:48:7-38
    But the last message received was:
    {
        "id": 2,
        "jsonrpc": "2.0",
        "result": []
    }
    Use -p '/Fix#3047/' to rerun this test only.

1 out of 6 tests failed (24.41s)
Test suite tests: FAIL
```

[Full build log](https://gist.github.com/hololeap/82ed57c88afd5a7d85c79e70f976a3ed)

This test passes outside of the sandbox.

**Comment by hololeap** [comment]
There are more seemingly-related test failures, this time from `ghcide`:

1. [`typecheck-all-parents-of-interest`](https://github.com/haskell/haskell-language-server/blob/ddb21a0c8d4e657c4b81ce250239bccf28fc9524/ghcide/test/exe/Main.hs#L698-L732)
   ```
   FAIL (2.13s)
         test/exe/Main.hs:724:
         Incorrect number of diagnostics for Uri {getUri = "file:///var/tmp/portage/dev-haskell/ghcide-1.8.0.0/temp/extra-dir-8862631353299/A.hs"}, expected [] but got [Diagnostic {_range = Range {_start = Positio
   n {_line = 5, _character = 4}, _end = Position {_line = 5, _character = 5}}, _severity = Just DsError, _code = Just (InR "-Wdeferred-type-errors"), _source = Just "typecheck", _message = "\8226 Couldn't match e
   xpected type \8216Int\8217 with actual type \8216Bool\8217\n\8226 In the expression: y\n  In an equation for \8216x\8217: x = y", _tags = Nothing, _relatedInformation = Nothing},Diagnostic {_range = Range {_sta
   rt = Position {_line = 5, _character = 4}, _end = Position {_line = 5, _character = 5}}, _severity = Just DsError, _code = Just (InR "-Wdeferred-type-errors"), _source = Just "typecheck", _message = "\8226 Coul
   dn't match expected type \8216Int\8217 with actual type \8216Bool\8217\n\8226 In the expression: y\n  In an equation for \8216x\8217: x = y", _tags = Nothing, _relatedInformation = Nothing}]
         Use -p '/typecheck-all-parents-of-interest/' to rerun this test only.
   ```

2. [`simple plugin`](https://github.com/haskell/haskell-language-server/blob/ddb21a0c8d4e657c4b81ce250239bccf28fc9524/ghcide/test/exe/Main.hs#L1191-L1200)
   ```
   FAIL (7.77s)
       test/exe/Main.hs:1196:
       Could not find (DsError,(9,15),"Variable not in scope: c",Nothing) in [Diagnostic {_range = Range {_start = Position {_line = 0, _character = 0}, _end = Position {_line = 1, _character = 0}}, _severity = Just DsError, _code = Nothing, _source = Just "cradle", _message = "Error when calling cabal exec -v0 -- ghc --print-libdir\n\ndieVerbatim: user error (cabal: '/usr/bin/curl' exited with an error:\ncurl: (6) Could not resolve host: hackage.haskell.org\n)\n\n", _tags = Nothing, _relatedInformation = Nothing}]
       Use -p '/simple plugin/' to rerun this test only.
   ```

3. [`notification handlers run in priority order`](https://github.com/haskell/haskell-language-server/blob/ddb21a0c8d4e657c4b81ce250239bccf28fc9524/ghcide/test/exe/Main.hs#L3194-L3213)
   ```
   FAIL
         Exception: Timed out waiting to receive a message from the server.
         Use -p '/notification handlers run in priority order/' to rerun this test only.
   ```

4. [`ignore-fatal-warning`](https://github.com/haskell/haskell-language-server/blob/ddb21a0c8d4e657c4b81ce250239bccf28fc9524/ghcide/test/exe/Main.hs#L2508-L2512)
   ```
   FAIL (9.15s)
           test/exe/Main.hs:2512:
           Got unexpected diagnostics for Uri {getUri = "file:///var/tmp/portage/dev-haskell/ghcide-1.8.0.0/temp/extra-dir-8862631353468/IgnoreFatal.hs"} got [Diagnostic {_range = Range {_start = Position {_line = 0, _character = 0}, _end = Position {_line = 1, _character = 0}}, _severity = Just DsError, _code = Nothing, _source = Just "cradle", _message = "Error when calling cabal exec -v0 -- ghc --print-libdir\n\ndieVerbatim: user error (cabal: '/usr/bin/curl' exited with an error:\ncurl: (6) Could not resolve host: hackage.haskell.org\n)\n\n", _tags = Nothing, _relatedInformation = Nothing}]
           Use -p '/ignore-fatal-warning/' to rerun this test only.
   ```

5. [`simple-multi-test`](https://github.com/haskell/haskell-language-server/blob/ddb21a0c8d4e657c4b81ce250239bccf28fc9524/ghcide/test/exe/Main.hs#L2526-L2538)
   ```
   FAIL (7.49s)
           test/exe/Main.hs:2532:
           A should typecheck
           Use -p '$0=="ghcide.cradle.multi.simple-multi-test"' to rerun this test only.
   ```

6. [`simple-multi-test2`](https://github.com/haskell/haskell-language-server/blob/ddb21a0c8d4e657c4b81ce250239bccf28fc9524/ghcide/test/exe/Main.hs#L2542-L2552)
   ```
   FAIL
           Exception: Timed out waiting to receive a message from the server.
           Last message received:
           {
               "jsonrpc": "2.0",
               "method": "window/logMessage",
               "params": {
                   "message": "Live bytes: 0.00MB Heap size: 0.00MB",
                   "type": 3
               }
           }
           Use -p '/simple-multi-test2/' to rerun this test only.
   ```

7. [`simple-multi-test3`](https://github.com/haskell/haskell-language-server/blob/ddb21a0c8d4e657c4b81ce250239bccf28fc9524/ghcide/test/exe/Main.hs#L2556-L2570)
   ```
   FAIL
           Exception: Timed out waiting to receive a message from the server.
           Last message received:
           {
               "jsonrpc": "2.0",
               "method": "window/logMessage",
               "params": {
                   "message": "Live bytes: 0.00MB Heap size: 0.00MB",
                   "type": 3
               }
           }
           Use -p '/simple-multi-test3/' to rerun this test only.
   ```

8. [`simple-multi-def-test`](https://github.com/haskell/haskell-language-server/blob/ddb21a0c8d4e657c4b81ce250239bccf28fc9524/ghcide/test/exe/Main.hs#L2574-L2588)
   ```
   FAIL
           Exception: Timed out waiting to receive a message from the server.
           Last message received:
           {
               "jsonrpc": "2.0",
               "method": "window/logMessage",
               "params": {
                   "message": "Live bytes: 0.00MB Heap size: 0.00MB",
                   "type": 3
               }
           }
           Use -p '/simple-multi-def-test/' to rerun this test only.
   ```

9. [`simple-subdirectory`](https://github.com/haskell/haskell-language-server/blob/ddb21a0c8d4e657c4b81ce250239bccf28fc9524/ghcide/test/exe/Main.hs#L2516-L2523)
   ```
   FAIL (5.22s)
           test/exe/Main.hs:2520:
           Could not find (DsWarning,(2,0),"Top-level binding",Nothing) in [Diagnostic {_range = Range {_start = Position {_line = 0, _character = 0}, _end = Position {_line = 1, _character = 0}}, _severity = Just DsError, _code = Nothing, _source = Just "cradle", _message = "Error when calling cabal exec -v0 -- ghc --print-libdir\n\ndieVerbatim: user error (cabal: '/usr/bin/curl' exited with an error:\ncurl: (6) Could not resolve host: hackage.haskell.org\n)\n\n", _tags = Nothing, _relatedInformation = Nothing}]
           Use -p '/simple-subdirectory/' to rerun this test only.
   ```

10. [`ghcide command line.works`](https://github.com/haskell/haskell-language-server/blob/ddb21a0c8d4e657c4b81ce250239bccf28fc9524/ghcide/test/exe/Main.hs#L2804-L2815)
   ```
     ghcide command line
       works:                                                                            FAIL (4.47s)
         test/exe/Main.hs:2814:
         expected: ExitSuccess
          but got: ExitFailure 1
         Use -p '/ghcide command line.works/' to rerun this test only.
   ```

```
13 out of 344 tests failed (2982.71s)
```

(Three tests fail regardless of the network-sandbox. See #3221.)

**Comment by pepeiborra** [comment]
The problem is that Tasty runs test in parallel by default, and most HLS test suites do not handle this well. Set TASTY_NUM_THREADS to 1

**Comment by michaelpj** [comment]
Did you try Pepe's suggestion? did it help?

**Comment by unhammer** [comment]
Possibly related, I was unable to start HLS in emacs without being online:

```
[stderr]  [2025-03-29T08:22:37Z INFO  emacs_lsp_booster::app] Running server "haskell-language-server-wrapper" "--lsp"
[stderr]  [2025-03-29T08:22:37Z INFO  emacs_lsp_booster::app] Will convert server json to bytecode! bytecode options: BytecodeOptions { object_type: Plist, null_value: Nil, false_value: Keyword("json-false") }
[stderr]  nil
[stderr]  No 'hie.yaml' found. Try to discover the project type!
[stderr]  Run entered for haskell-language-server-wrapper(haskell-language-server-wrapper) Version 2.0.0.0 x86_64 ghc-9.2.8
[stderr]  Current directory: /home/me/foo
[stderr]  Operating system: linux
[stderr]  Arguments: ["--lsp"]
[stderr]  Cradle directory: /home/me/foo
[stderr]  Cradle type: Cabal
[stderr]
[stderr]
[stderr]  Tool versions found on the $PATH
[stderr]  cabal:          3.12.1.0
[stderr]  stack:          3.1.1
[stderr]  ghc:            9.2.8
[stderr]
[stderr]
[stderr]  Consulting the cradle to get project GHC version...
[stderr]  2025-03-29T08:22:38.402607Z | Debug | executing command: cabal exec -v0 -- ghc --print-libdir
[stderr]  nil
[stderr]  Failed to find the GHC version of this Cabal project.
[stderr]  Error when calling cabal exec -v0 -- ghc --print-libdir
[stderr]
[stderr]  dieVerbatim: user error (Error: cabal: '/usr/bin/curl' exited with an error:
[stderr]  curl: (6) Could not resolve host: hackage.haskell.org
[stderr]  )
```

then I connected (now in the terminal) and saw that it complained about cabal not being updated:

```sh
$ haskell-language-server-wrapper
No 'hie.yaml' found. Try to discover the project type!
Run entered for haskell-language-server-wrapper(haskell-language-server-wrapper) Version 2.0.0.0 x86_64 ghc-9.2.8
Current directory: /home/me/foo
Operating system: linux
Arguments: []
Cradle directory: /home/me/foo
Cradle type: Cabal

Tool versions found on the $PATH
cabal:          3.12.1.0
stack:          3.1.1
ghc:            9.2.8


Consulting the cradle to get project GHC version...
2025-03-29T08:24:48.403936Z | Debug | executing command: cabal exec -v0 -- ghc --print-libdir
Failed to find the GHC version of this Cabal project.
Error when calling cabal exec -v0 -- ghc --print-libdir

Error: [Cabal-7160]
The package list for 'hackage.haskell.org' does not exist. Run 'cabal update' to download it.
```

I don't understand why it needs to do that.

Anyway, I did `cabal update` and now I am able to use HLS in airplane mode again.

(Yes I should update HLS, I haven't found time to test on the newest, but hopefully this helps anyone looking for the same error message.)
